[jit] serializing function calls

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23799 [jit] serializing function calls**

Before, we inlined as part of the initial IR generation process, which
has a few disadvantages:

1. It loses information about what nodes came from which function/method
calls. Other parties who want to implement transformations on the
function/module level don't have a reliable way of doing so.
2. It duplicates a ton of code if we are inlining the same
function/method a tons of times.

After this PR: inline is deferred to the optimization stage, so
optimizations that rely on inlining will still work. But things get
serialized with the function/method calls in.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23799

Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
[wip] serializing function calls

suo committed:
Update on "[wip] serializing function calls"

suo committed:
Update on "[wip] serializing function calls"


Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[wip] serializing function calls"


Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[wip] serializing function calls"


Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[wip] serializing function calls"


Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[wip] serializing function calls"


Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[wip] serializing function calls"


Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[wip] serializing function calls"


Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[wip] serializing function calls"


Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[wip] serializing function calls"


Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo commented:
heads up: @ZolotukhinM

suo committed:
Update on "[jit] serializing function calls"


Before, we inlined as part of the initial IR generation process, which
has a few disadvantages:

1. It loses information about what nodes came from which function/method
calls. Other parties who want to implement transformations on the
function/module level don't have a reliable way of doing so.
2. It duplicates a ton of code if we are inlining the same
function/method a tons of times.

After this PR: inline is deferred to the optimization stage, so
optimizations that rely on inlining will still work. But things get
serialized with the function/method calls in.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23799

Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[jit] serializing function calls"


Before, we inlined as part of the initial IR generation process, which
has a few disadvantages:

1. It loses information about what nodes came from which function/method
calls. Other parties who want to implement transformations on the
function/module level don't have a reliable way of doing so.
2. It duplicates a ton of code if we are inlining the same
function/method a tons of times.

After this PR: inline is deferred to the optimization stage, so
optimizations that rely on inlining will still work. But things get
serialized with the function/method calls in.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23799

Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[jit] serializing function calls"


Before, we inlined as part of the initial IR generation process, which
has a few disadvantages:

1. It loses information about what nodes came from which function/method
calls. Other parties who want to implement transformations on the
function/module level don't have a reliable way of doing so.
2. It duplicates a ton of code if we are inlining the same
function/method a tons of times.

After this PR: inline is deferred to the optimization stage, so
optimizations that rely on inlining will still work. But things get
serialized with the function/method calls in.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23799

Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[jit] serializing function calls"


Before, we inlined as part of the initial IR generation process, which
has a few disadvantages:

1. It loses information about what nodes came from which function/method
calls. Other parties who want to implement transformations on the
function/module level don't have a reliable way of doing so.
2. It duplicates a ton of code if we are inlining the same
function/method a tons of times.

After this PR: inline is deferred to the optimization stage, so
optimizations that rely on inlining will still work. But things get
serialized with the function/method calls in.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23799

Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

zdevito commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -61,6 +61,10 @@ struct TORCH_API ClassNamespaceValue : public SugaredValue {
       }
     }
 
+    if (auto fn = cu_.find_function(fullName)) {
```
What are these? Can you add a comment?

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -257,8 +257,13 @@ struct PythonPrintPass {
     }
-    if (std::find(deps_table_.cbegin(), deps_table_.cend(), type) ==
-        deps_table_.cend()) {
+    // Need to do actual equality comparison, not a pointer equality.
```
Can you make the comment more helpful? I can read that from the code, but I want to know why

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -949,6 +954,10 @@ struct PythonPrintPass {
         stmt << "uninitialized(" << node->output()->type()->python_str() << ")";
       } break;
       case prim::Constant: {
+        if (node->outputs().size() == 1 &&
```
It is very confusing to have a function printRHS that always prints, except for function constants. Can this be refactored? It will trip up someone trying to understand this code.

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -1039,6 +1048,33 @@ struct PythonPrintPass {
         }
       } break;
+      case prim::CallFunction: {
+        const auto& fn = node->inputs().at(0)->type()->expect<FunctionType>();
```
Why not just call print rhs Here?

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -1039,6 +1048,33 @@ struct PythonPrintPass {
+            method->qualname() ==
+            QualifiedName(selfType->name()->qualifiedName(), methodName));
+
+        stmt << useOf(self) << "." << methodName << "(";
```
Is there any operator we print inline that is lower precedence than `.`? If so we need to put parentheses around self.

suo replied:
I guess might as well just to be safe

zdevito reviewed (approved):
Looks good! Just some minor clarity notes.

suo committed:
Update on "[jit] serializing function calls"


Before, we inlined as part of the initial IR generation process, which
has a few disadvantages:

1. It loses information about what nodes came from which function/method
calls. Other parties who want to implement transformations on the
function/module level don't have a reliable way of doing so.
2. It duplicates a ton of code if we are inlining the same
function/method a tons of times.

After this PR: inline is deferred to the optimization stage, so
optimizations that rely on inlining will still work. But things get
serialized with the function/method calls in.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23799

Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)

suo committed:
Update on "[jit] serializing function calls"


Before, we inlined as part of the initial IR generation process, which
has a few disadvantages:

1. It loses information about what nodes came from which function/method
calls. Other parties who want to implement transformations on the
function/module level don't have a reliable way of doing so.
2. It duplicates a ton of code if we are inlining the same
function/method a tons of times.

After this PR: inline is deferred to the optimization stage, so
optimizations that rely on inlining will still work. But things get
serialized with the function/method calls in.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23799

Differential Revision: [D16652819](https://our.internmc.facebook.com/intern/diff/D16652819)
